Orm/Xtensive.Orm/Orm/Upgrade/Internals/HintGenerator.cs

```diff
@@ -218,64 +219,73 @@ private void GenerateCopyColumnHint(CopyFieldHint hint)
-        .ForEach(t => newTypes.Add($"{t.MappingDatabase}.{t.MappingSchema}.{t.MappingName}", t));
-
-      foreach (var rType in removedTypes) {
-        var rTypeIdentifier = $"{rType.MappingDatabase}.{rType.MappingSchema}.{rType.MappingName}";
-        if (!suspiciousTypes.Contains(rType)
-          || !newTypes.TryGetValue(rTypeIdentifier, out var conflictedNewType)) {
-          continue;
-        }
-        schemaHints.Add(new RecreateTableHint(GetTablePath(rType)));
-      }
-    }
-
-    private void GenerateRecordCleanupHints(List<StoredTypeInfo> removedTypes, bool isMovedToAnotherHierarchy)
+    private void GenerateRecordCleanupHints(List<StoredTypeInfo> removedTypes,
+      Dictionary<StoredTypeInfo, StoredTypeInfo> conflictsByTable, bool isMovedToAnotherHierarchy)
     {
       if (!isMovedToAnotherHierarchy) {
         removedTypes.ForEach(GenerateCleanupByForeignKeyHints);
       }
-      removedTypes.ForEach(type => GenerateCleanupByPrimaryKeyHints(type, isMovedToAnotherHierarchy));
+      removedTypes.ForEach(type => GenerateCleanupByPrimaryKeyHints(type, isMovedToAnotherHierarchy, conflictsByTable.ContainsKey(type.Hierarchy.Root)));
     }
 
-    private void GenerateCleanupByPrimaryKeyHints(StoredTypeInfo removedType, bool isMovedToAnotherHierarchy)
+    private void GenerateCleanupByPrimaryKeyHints(StoredTypeInfo removedType, bool isMovedToAnotherHierarchy, bool conflictsWithNewType)
     {
       var hierarchy = removedType.Hierarchy;
       switch (hierarchy.InheritanceSchema) {
-        case InheritanceSchema.ClassTable:
-          IEnumerable<StoredTypeInfo> typesToProcess;
-          typesToProcess = !isMovedToAnotherHierarchy
-            ? EnumerableUtils.One(removedType).Concat(removedType.AllAncestors)
-            : removedType.AllAncestors;
-          foreach (var type in typesToProcess) {
-            var identities1 = new[] { new IdentityPair(
-              GetColumnPath(type, GetTypeIdMappingName(type)),
-              removedType.TypeId.ToString(),
-              true) };
-            schemaHints.Add(
-              new DeleteDataHint(GetTablePath(type), identities1, isMovedToAnotherHierarchy));
+        case InheritanceSchema.ClassTable: {
+          if (!conflictsWithNewType) {
+            var typesToProcess = !isMovedToAnotherHierarchy
+              ? EnumerableUtils.One(removedType).Concat(removedType.AllAncestors)
+              : (IEnumerable<StoredTypeInfo>) removedType.AllAncestors;
+            foreach (var type in typesToProcess) {
+              var identities1 = new[] { new IdentityPair(
+                GetColumnPath(type, GetTypeIdMappingName(type)),
+                removedType.TypeId.ToString(),
+                true)
+              };
+              schemaHints.Add(
+                new DeleteDataHint(GetTablePath(type), identities1, isMovedToAnotherHierarchy, conflictsWithNewType));
+            }
+          }
+          else if (isMovedToAnotherHierarchy) {
```

Comment:
Why this condition is different from the same one in the other two cases?

Reply:
this is mistake